Error reporting on misspelled block args

bazsi opened the pull request:
This branch makes the __VARARGS__ functionality in SCL blocks explicit, as currently it is very easy to mis-spell an argument, which is not logged in any way. It changes the syntax slightly:

```
block source something(foo() bar() ...) {
};
```

The ... at the end of the argument list tells syslog-ng that this macro indeed wants to accept __VARARGS__, thus any name-value pair can be passed without validation.

bazsi commented:
btw, this depends on the improve-error-handling branch, so that needs to go in first, I just wanted to throw this into the mix.

kira-syslogng commented:
failure

kira-syslogng commented:
failure

kira-syslogng commented:
success

bazsi commented:
rebased against master, remove WIP tag, this should be ok to review.

kira-syslogng commented:
success

furiel commented on lib/cfg-lexer.c:
```diff
@@ -946,18 +946,22 @@ cfg_lexer_lex(CfgLexer *self, YYSTYPE *yylval, YYLTYPE *yylloc)
+                success = cfg_lexer_include_buffer(self, buf, result->str, result->len);
+              g_string_free(result, TRUE);
+
               goto relex;
```
The original code conditionally goto-ed to relex, when `g_lexer_include_buffer_without_backtick_substitution` and `cfg_lexer_include_buffer` returned successful. Can we skip to check the return values now?

furiel commented on lib/cfg-block.c:
```diff
@@ -61,6 +61,50 @@ cfg_block_format_name(CfgBlockGenerator *s, gchar *buf, gsize buf_len)
+  CfgArgs *defs = ((gpointer *) user_data)[0];
+  const gchar *reference = ((gpointer *) user_data)[1];
+  gpointer *problem_found = &((gpointer *) user_data)[2];
+
```
As I see problem_found is write-only now? What is the purpose? Is an early return missing in this function?

furiel replied:
Ok I see it is used on line 100 with `if (validate_params[2])`. The early early return might still be useful.

bazsi replied:
what do you mean on early return? the entire function is just one if statement, the iteration happens in the foreach which uses this as a callback. You think that the return would make this more readable?

furiel replied:
I was thinking if the previous iteration ran into an error, then we could skip the execution of this function if we return when problem_found is true.

Maybe it would be better use g_hash_table_find instead of g_hash_table_foreach to avoid unnecessary calls. With g_hash_table_find, probably we could eliminate the problem_found variable.

Though I agree this is not a performance heavy path, so we can stick with anything that is more readable than optimal.

furiel commented on lib/cfg-block.c:
```diff
@@ -61,6 +61,50 @@ cfg_block_format_name(CfgBlockGenerator *s, gchar *buf, gsize buf_len)
+static gboolean
+_validate_args(CfgArgs *self, CfgArgs *defs, const gchar *reference)
+{
+  gpointer validate_params[] = { defs, (gchar *) reference, NULL };
```
Please use FALSE instead of NULL.

bazsi replied:
I've changed this code to pass a gboolean* as third parameter so that I can name the returned value, which makes the code easier to read. I'll push this in a moment.

furiel commented on lib/cfg-block.c:
```diff
@@ -53,11 +53,11 @@ cfg_block_format_name(CfgBlockGenerator *s, gchar *buf, gsize buf_len)
   CfgBlock *self = (CfgBlock *) s;
 
-  g_snprintf(buf, buf_len, "block %s %s (at %s:%d:%d)",
+  g_snprintf(buf, buf_len, "block %s %s() at %s:%d",
```
() is left for separator purposes?

bazsi replied:
In the documentation and log messages we use a convention that a paren is there at the end of the driver, e.g. unix-stream() instead of unix-stream.

This just follows that convention.

kira-syslogng commented:
success

bazsi commented:
I think I have now addressed the concerns, except for the "early return" that @furiel requested and I didn't completely understand. I've also rebased to master.

One thing that occurred me is if we want to do this selectively on version number, @fekete-robert @czanik what do you think?

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build SUCCESS

gaborznagy commented on lib/cfg-block.c:
```diff
@@ -61,6 +61,50 @@ cfg_block_format_name(CfgBlockGenerator *s, gchar *buf, gsize buf_len)
+    {
+      if (cfg_args_is_accepting_varargs(defs))
+        {
+          msg_verbose("Unknown argument, adding it to __VARARGS__",
```
My understanding is that here we cannot differentiate between valid driver options and invalid ones.
We can only know if the option is defined in the scl block or not.

I wonder how this part helps in debugging as if the user configures a lot of options and an invalid option too, he/she will see the parsing error anyway.

bazsi replied:
This makes it possible to see that an argument was indeed not defined explicitly, so even if you have '...' as argument, you'll see (in debug mode) that an argument is using the wildcard mode.

I agree that it could trigger an error later on (when pasting __VARARGS__ into somewhere), but this message can be a useful when __VARARGS__ is pasted to somewhere where it is not actually validated.

Maybe this should be a debug message, instead of a verbose, I'll do that before merging.

kira-syslogng commented:
Build FAILURE

furiel commented:
@kira-syslogng retest this please

kira-syslogng commented:
Build SUCCESS
